### PR TITLE
Issue 6712: Table Segment index recovery tool

### LINKED ISSUE #6712: Corrupted BTreeIndex footer found after a segment container recovery when cache is full
**Describe the bug**
In a longevity workload (heavy workload mixing transactions and regular events), an issue was found after 24 days of uninterrupted operation. The sequence of steps:
1. All segment stores were working normally, until one of them started to show symptoms of cache being full:
```
2022-04-18 01:24:35,584 2138732271 [core-27] INFO  i.p.s.server.CacheManager - CacheManager: Cache full. Forcing cache policy.
...
2022-04-18 01:24:36,091 2138732778 [core-1] WARN  i.p.s.s.logs.MemoryStateUpdater - Not adding operation 'StreamSegmentAppendOperation: SequenceNumber = 866200873, SegmentId = 476901, Offset = 1784640543791, Length = 27594, Attributes = 2' to ReadIndex because the Cache is full.
2022-04-18 01:24:36,093 2138732780 [core-8] INFO  i.p.s.server.CacheManager - CacheManager: Cache full. Forcing cache policy.
2022-04-18 01:24:36,093 2138732780 [core-29] ERROR i.p.s.s.logs.OperationProcessor - OperationProcessor[1]: MemoryStateUpdater.process failure.
io.pravega.segmentstore.server.ServiceHaltException: Unable to add operation 'StreamSegmentAppendOperation: SequenceNumber = 866200874, SegmentId = 476901, Offset = 1784640571385, Length = 45651, Attributes = 2' to ReadIndex.
	at io.pravega.segmentstore.server.logs.MemoryStateUpdater.addToReadIndex(MemoryStateUpdater.java:276)
	at io.pravega.segmentstore.server.logs.MemoryStateUpdater.process(MemoryStateUpdater.java:196)
	at io.pravega.segmentstore.server.logs.MemoryStateUpdater.process(MemoryStateUpdater.java:158)
	at io.pravega.segmentstore.server.logs.OperationProcessor.processCommits(OperationProcessor.java:479)
	at io.pravega.segmentstore.server.logs.OperationProcessor.handleProcessCommits(OperationProcessor.java:188)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:189)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.IllegalArgumentException: The given range of bytes (Offset=1784640571385) does not start right after the last appended offset (1784640543790).
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:385)
	at io.pravega.segmentstore.server.reading.StreamSegmentReadIndex.append(StreamSegmentReadIndex.java:399)
```
2. At this point, containers in this segment store were restarting. For the sake of conciseness, we can see that recoveries of Segment Container 32 complete normally, including the recovery of the SegmentAttributeIndex of Segment `476467211`:
```
2022-04-18 01:24:53,013 2138749700 [core-2] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[32] Recovery started.
...
2022-04-18 01:24:37,109 2138733796 [core-17] INFO  i.p.s.server.CacheManager - CacheManager Unregistered ReadIndex[32-476467211] (_system/_tables/completedTransactionsBatch-24).
...
2022-04-18 01:24:55,916 2138752603 [core-2] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[32] Recovery completed. Epoch = 9501, Items Recovered = 6575, Time = 2902ms.
...
2022-04-18 01:25:13,514 2138770201 [core-18] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[32-476467211]: Initialized. StorageLength = 83189810, Sealed = false.
2022-04-18 01:25:13,514 2138770201 [core-18] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[32-476467211]: State changed from NotInitialized to Writing.
2022-04-18 01:25:13,550 2138770237 [core-9] INFO  i.p.common.util.btree.BTreeIndex - AttributeIndex[32-476467211]: Initialized. State = Length = 12619552104, RootOffset = 12619519506, RootLength = 32586, Stats = null.
2022-04-18 01:25:13,550 2138770237 [core-9] INFO  i.p.s.server.CacheManager - CacheManager Registered AttributeIndex[32-476467211].
```
3. Note that this SegmentAttributeIndex (`476467211`) gets initialized multiple times. Also, we can observe the length progressing, probably due to the progress of segment attributes being written to LTS:
```
2022-04-18 01:32:01,462 2139178149 [core-30] INFO  i.p.common.util.btree.BTreeIndex - AttributeIndex[32-476467211]: Initialized. State = Length = 12700724422, RootOffset = 12700724328, RootLength = 82, Stats = null.
2022-04-18 01:32:01,462 2139178149 [core-30] INFO  i.p.s.server.CacheManager - CacheManager Registered AttributeIndex[32-476467211].
...
2022-04-18 01:38:11,425 2139548112 [core-30] INFO  i.p.common.util.btree.BTreeIndex - AttributeIndex[32-476467211]: Initialized. State = Length = 12778338702, RootOffset = 12778338608, RootLength = 82, Stats = null.
2022-04-18 01:38:11,425 2139548112 [core-30] INFO  i.p.s.server.CacheManager - CacheManager Registered AttributeIndex[32-476467211].
```

4. However, after 20 minutes in this situation, we observe that there is a `DataCorruptionException` initializing the SegmentAttributeIndex for Segment `476467211` in Segment Container 32:

```
2022-04-18 01:44:13,832 2139910519 [core-29] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[32-476467211]: Initialized. StorageLength = 84588130, Sealed = false.
2022-04-18 01:44:13,832 2139910519 [core-29] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[32-476467211]: State changed from NotInitialized to Writing.
...
2022-04-18 01:44:13,934 2139910621 [core-25] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[32]: Iteration[6].CriticalError.
io.pravega.segmentstore.server.DataCorruptionException: BTreeIndex operation failed. Index corrupted.
	at io.pravega.segmentstore.server.attributes.SegmentAttributeBTreeIndex.handleIndexOperationException(SegmentAttributeBTreeIndex.java:808)
	at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:189)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.common.util.IllegalDataFormatException: [AttributeIndex[32-476467211]] Wrong footer information. RootPage Offset (6805637576853986517) + Length (-1140654080) exceeds Footer Offset (12845447554).
	at io.pravega.common.util.btree.BTreeIndex.initialize(BTreeIndex.java:235)
	at io.pravega.common.util.btree.BTreeIndex.lambda$initialize$0(BTreeIndex.java:213)
	... 7 common frames omitted
```

5. When inspecting the footer contents from the storage (i.e., reading the last 30 bytes of that BTreeIndex on LTS), we see that the wrong lengths seen in the above error can be also found there:
```
7500802929515823104,  1207,  8556285034968135232,  6805637576853986517,  -1140654080
```
This implies that the last write for the footer may be writing wrong/corrupted data, which impedes the initialization of the BTreeIndex in the next recovery.

**To Reproduce**
Hard to do, the code has been there for years and it is the first time we find this issue. Probably related to write the footer of a BTree under complete cache full situation, but we still need to reproduce it via tests.

**Screenshots**
n/a

**Additional information**
Apart from finding the root cause, it would be interesting to see if we can come up with a repair process for this particular BTreeIndex.
Also, it could also be the case that the LTS storage could have corrupted the data, but this looks more unlikely.

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**  
Adds a new Admin CLI command plus the associated instructions to repair a corrupted Table Segment index in storage from the main Table Segment data.

**Purpose of the change**  
Fixes #6712.

**What the code does**  
This PR tries to approach fixing #6712 from a repair perspective. In the original issue, we observed that the footer of the last index chunk for a Table Segment was corrupted. This could be due to a rare bug in Pravega, or even to a problem in the Storage system itself. In any case, we need tools to deal with such a situation in which a Table Segment is corrupted. To this end, this PR does the following:
- New Admin CLI `dataRecovery tableSegment-recovery`:
  - The assumption of this command is that the main Table Segment data (i.e., storage chunks where PUT/DELETE operations are stored) are available to reconstruct the index.
  - One input parameter for the command is the location (i.e., local folder) with the data chunks related to the Table Segment. This means that the user should be able to access this data. The second parameter is the name of the new Table Segment to be created to reconstruct the original one's index.
  - The command works in 3 main steps: i) ordered scan of the Table Segment data chunks to parse the `TableEntry` operations stored, ii) start a new local Pravega instance (with SLTS Storage enabled), iii) process all the Table Segment operations related to the parsed `TableEntry` and flush data to storage.
- With the above, the user should be able to have a local reconstruction of index for the original Table Segment data.
- After that, we also provide the instructions step-by-step on how to replace the newly reconstructed index chunks in the original cluster (container and storage metadata changes).

**How to verify it**  
New test added to test the new command.
Manual testing in progress.

### CODE REVIEW

**Comment by RaulGracia** [comment]
@sachin-j-joshi @abhinb please, can you have a look?

**Comment by abhinb** [comment]
At times we do see the below stacktrace generated while shutting down the in-process pravega which might confuse the user if the command has worked or not. 

We do see that the command could flush the output chunks.
`Flushed the Segment Container with containerId 0 to Storage.`

Post that is when we can see the below trace:-
> com.google.common.util.concurrent.ListenerCallQueue$PerListenerQueue run
SEVERE: Exception while executing callback: io.pravega.common.concurrent.Services$ShutdownListener@313c465f terminated({from = STOPPING})
java.util.concurrent.RejectedExecutionException: Task ThreadPoolScheduledExecutorService.ScheduledRunnable(id=1414392, isDelayed=false, scheduledTimeNanos=0, task=java.util.concurrent.Executors$RunnableAdapter@2e79c08[Wrapped task = io.pravega.segmentstore.server.reading.StreamSegmentReadIndex$$Lambda$2170/0x00000008008d4840@9253a1], future=java.util.concurrent.CompletableFuture@35ee855a[Not completed]) rejected from java.util.concurrent.ThreadPoolExecutor@f058c6b[Shutting down, pool size = 22, active threads = 1, queued tasks = 0, completed tasks = 1248059]
        at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.reject(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.execute(Unknown Source)
        at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService.execute(ThreadPoolScheduledExecutorService.java:243)
        at io.pravega.segmentstore.server.reading.StreamSegmentReadIndex.close(StreamSegmentReadIndex.java:175)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.closeIndex(ContainerReadIndex.java:425)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.lambda$closeAllIndices$0(ContainerReadIndex.java:434)
        at java.base/java.util.ArrayList.forEach(Unknown Source)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.closeAllIndices(ContainerReadIndex.java:434)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.close(ContainerReadIndex.java:115)
        at io.pravega.segmentstore.server.containers.StreamSegmentContainer.close(StreamSegmentContainer.java:270)
        at io.pravega.segmentstore.server.store.StreamSegmentContainerRegistry.unregisterContainer(StreamSegmentContainerRegistry.java:203)
        at io.pravega.segmentstore.server.store.StreamSegmentContainerRegistry.lambda$startContainerInternal$1(StreamSegmentContainerRegistry.java:186)
        at io.pravega.common.concurrent.Services$ShutdownListener.terminated(Services.java:126)
        at com.google.common.util.concurrent.AbstractService$3.call(AbstractService.java:95)
        at com.google.common.util.concurrent.AbstractService$3.call(AbstractService.java:92)
        at com.google.common.util.concurrent.ListenerCallQueue$PerListenerQueue.run(ListenerCallQueue.java:205)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:193)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)

**Comment by abhinb** [comment]
should the sizes of the rolled over chunks created by this command be the same as that on the affected cluster to maintain consistency with chunk sizes? In most cases the size is going to be the default 32M i suppose right?. But just in case it differs.

File: cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+                // Add the operation to the list of operations to replay later on (PUT or DELETE).
+                tableSegmentOperations.add(valueBytes.length == 0 ?
+                        new DeleteOperation(TableSegmentEntry.versioned(keyBytes, valueBytes, header.getEntryVersion())) :
+                        new PutOperation(TableSegmentEntry.versioned(keyBytes, valueBytes, header.getEntryVersion())));
+                // Full entry read, so reset unprocessed bytes.
+                Preconditions.checkState(unprocessedBytesFromLastChunk < totalEntryLength, "Some bytes are missing to process.");
+                unprocessedBytesFromLastChunk = 0;
+                processedBytes += totalEntryLength;
+                // Show the progress of scanning Table Segment operations.
+                if (tableSegmentOperations.size() % 100 == 0) {
```
Comment (abhinb) [inline cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java:211]: should we print the name of the chunk here. else it just looks like the percentage has fallen when we move to the next chunk file.

Comment (RaulGracia) [reply cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java:217]: I have added the print file name before the execution of this method, so the progress related to that file.

**Comment by RaulGracia** [comment]
@abhinb the message you posted may related to the shutdown of the local Pravega instance. I don't know if I can do much to do not show it. What I have added is a new print specifying that the command has been executed correctly.
I have also added the possibility to parameterize the rolling length of chunks.

**Review by abhinb — APPROVED** [review]

**Review by kuldeepk3 — APPROVED** [review]

File: cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    }
+
+    /**
+     * Start a local Pravega cluster with SLTS enabled, so we can use the resulting Table Segment index chunks.
+     */
+    private void setupStorageEnabledLocalPravegaCluster() throws Exception {
+        output("Storage directory for this cluster is: " + this.pravegaStorageDir);
+        FileSystemStorageConfig adapterConfig = FileSystemStorageConfig.builder()
+                .with(FileSystemStorageConfig.ROOT, this.pravegaStorageDir.toString())
+                .with(FileSystemStorageConfig.REPLACE_ENABLED, true)
```
Comment (sachin-j-joshi) [inline cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java:148]: REPLACE_ENABLED is not used in SLTS

**Review by sachin-j-joshi — COMMENTED** [review]
One concern I have is that there are edge cases like during failover or zombie instances and other LTS failures. 

In such cases, we might end up writing extra junk data at the end of chunks or even extra junk chunks that should not be included in valid data.  This can also happen when there are zombie instances still writing to old chunks while at the same time new instance has started new chunks.

If we blindly enumerate all chunks on LTS and process then it might actually end up processing irrelevant/incorrect data.

Now, the SLTS segment/chunk metadata has correct start and end offsets for all valid data. 

One solution is to use this method to retrieve list of valid chunks directly from SLTS https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/UtilsWrapper.java#L217

File: cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java (hunk 2 above)
Comment (RaulGracia) [reply cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java]: Ok, removed.

File: cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                .build();
+        StorageFactory storageFactory = new FileSystemSimpleStorageFactory(ChunkedSegmentStorageConfig.DEFAULT_CONFIG, adapterConfig, this.executor);
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java:149]: I have reviewed the configuration. From the related parameters, I don't see anything dangerous from the default values for `maxIndexedSegments`, `maxIndexedChunksPerSegment`, `maxIndexedChunks`. In any case, if we ever hit any issue is just a config that can be change if ever needed.

**Comment by RaulGracia** [comment]
Thanks @sachin-j-joshi, your comment about the source of the primary chunks for the tool may be relevant. However, this tool just assumes that the right chunks are provided as input, so the index can be reconstructed. How to correctly gather the primary data chunks is a different topic, so your observations would be important to be included in #6760. cc/ @abhinb

**Pull request merged**